docs/content/docs/deployment/speculative_execution.md

```diff
@@ -0,0 +1,112 @@
+The multiplier to calculate the slow tasks detection baseline. Given that the parallelism is N and the ratio is R, define T as the median of the first `N*R` finished tasks' execution time. The baseline will be `T*M`, where M is the multiplier of the baseline.
+- [`slow-task-detector.execution-time.baseline-ratio`]({{< ref "docs/deployment/config" >}}#slow-task-detector-execution-time-baseline-ratio): 
+The finished execution ratio threshold to calculate the slow tasks detection baseline. Given that the parallelism is N and the ratio is R, define T as the median of the first `N*R` finished tasks' execution time. The baseline will be `T*M`, where M is the multiplier of the baseline.
+
+### Enable Sources for Speculative Execution
+If your job uses a custom {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/Source.java" name="Source" >}}, 
+and the source uses custom {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SourceEvent.java" name="SourceEvent" >}},
+you need to change the {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SplitEnumerator.java" name="SplitEnumerator" >}} 
+of that source to implement {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SupportsHandleExecutionAttemptSourceEvent.java" name="SupportsHandleExecutionAttemptSourceEvent" >}} 
+interface.
+```java
+public interface SupportsHandleExecutionAttemptSourceEvent {
+    void handleSourceEvent(int subtaskId, int attemptNumber, SourceEvent sourceEvent);
+}
+```
+This means the SplitEnumerator should be aware of the attempt which sends the event. Otherwise, exceptions 
+will happen when the job manager receives a source event from the tasks and lead to job failures.
+
+No extra change is required for other sources to work with speculative execution, including 
+{{< gh_link file="/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/SourceFunction.java" name="SourceFunction sources" >}}, 
+{{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/common/io/InputFormat.java" name="InputFormat sources" >}}, 
+and {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/Source.java" name="new sources" >}}. 
+All the source connectors offered by Apache Flink can work with speculative execution.
+
+## Checking the Effectiveness of Speculative Execution
+After enabling speculative execution, when there are slow tasks that trigger speculative execution, 
+the web UI will show the speculative attempts on the `SubTasks` tab of vertices on the job page. The web UI 
+also shows the blocked taskmanagers on the Flink cluster `Overview` and `Task Managers` pages.
+
+You can also check below metrics to see the effectiveness of speculative execution:
+- **numSlowExecutionVertices**: Number of slow execution vertices at the moment.
```

Comment:
Maybe we should add metric description in https://nightlies.apache.org/flink/flink-docs-master/docs/ops/metrics, and link it here.

Reply:
Yes, we should do that.